pilot/pkg/model/mesh.go

```diff
@@ -0,0 +1,785 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package model provides Mesh which is a platform independent
+// abstraction used by Controllers to synchronize the list of endpoints
+// used by the pilot with those available in the platform specific
+// registry. It implements all the necessary logic that's used for service
+// discovery based on routing rules and endpoint subsets.
+// Typical Usage:
+//
+//   import "istio.io/istio/pilot/model"
+//
+//   type MyPlatformController struct {
+//     Mesh
+//   }
+//   ...
+//   pc := MyPlatformController{model.NewMesh()}
+//   ...
+//   var allEndpoints []*model.ServiceInstances
+//   nativeEndpoints = buildYourPlatformServiceEndpointList()
+//   meshEndpoints := make([]*Endpoint, len(nativeEndpoints))
+//	 for idx, nativeEp := range nativeEndpoints {
+//     // Create mesh endpoint from relevant values of nativeEp
+//	   meshEndpoints[idx] = NewEndpoint(......)
+//   }
+//   pc.Reconcile(meshEndpoints)
+//
+package model
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"net"
+	"reflect"
+	"regexp"
+	"sort"
+	"strconv"
+	"sync"
+
+	xdsapi "github.com/envoyproxy/go-control-plane/api"
+	types "github.com/gogo/protobuf/types"
+	multierror "github.com/hashicorp/go-multierror"
+	route "istio.io/api/routing/v1alpha2"
+	"istio.io/istio/pkg/log"
+)
+
+const (
+	// The internal filter name for attributes that should be used for
+	// constructing Istio attributes on the Endpoint in
+	// <a href='https://goo.gl/mFzjpp'>reverse DNS form</a>.
+	istioConfigFilter = "io.istio.istio-config"
+
+	// Key Istio attribute names for mapping endpoints to subsets.
+	// For details, please see
+	// https://istio.io/docs/reference/config/mixer/attribute-vocabulary.html
+
+	// UID is the attribute name for the mesh unique, platform-specific
+	// unique identifier for the server instance of the destination service. Mesh
+	// uses the label to uqniquely identify the endpoint when it receives updates
+	// from the Controller. Example: kubernetes://my-svc-234443-5sffe.my-namespace
+	// No two instances running anywhere in the mesh can have the same value for
+	// UID.
+	UID destinationAttribute = "destination.uid"
+
+	// SERVICE represents the fully qualified __CANONOCAL__ name of the Istio
+	// service, ex: "my-svc.my-namespace.svc.cluster.local".  Mesh uses
+	// this label to identify endpoints belonging to the same service for XDS queries
+	// involving service names. The attribute value is expected to be mesh unique,
+	// meaning services having this name are semantically understood to be the same
+	// modulo version and other labels, irrespective of the platform under which
+	// they are run, the clusters or availability zones they run in.
+	SERVICE destinationAttribute = "destination.service"
```

Comment:
Go style is to capitalize first letter only.

Reply:
Ah bad habits carried over from C++ and Java. I ask for forgiveness. In any case, given the 3 other rounds of review I was hoping to get early feedback, so didn't run linter on this file. Sorry for any heartburn.